src/theory/sep/theory_sep.cpp

```diff
@@ -1744,105 +1747,128 @@ bool TheorySep::areDisequal( Node a, Node b ){
+        // two positive pto
+        if (!areEqual(pval, qval))
         {
-          if( areEqual( atom[1], ei_n ) ){
-            addPto( ei, ei_n, atom, false );
+          std::vector<Node> exp;
+          if (plbl != qlbl)
+          {
+            Assert(areEqual(plbl, qlbl));
+            exp.push_back(plbl.eqNode(qlbl));
+          }
+          exp.push_back(p);
+          exp.push_back(q);
+          // enforces injectiveness of pto
+          //  (pto x y) ^ (pto y w) ^ x = y => y = w
+          Node concn = pval.eqNode(qval);
+          Trace("sep-pto") << "prop pos/pos: " << concn << " by " << exp
+                           << std::endl;
+          sendLemma(exp, concn, InferenceId::SEP_PTO_PROP);
+          // Don't need to add this if the labels are identical. This is an
+          // optimization to minimize the size of the pto list
+          if (plbl == qlbl)
+          {
+            ret = false;
           }
         }
       }
-    }
-    //we have now processed all pending negated pto
-    ei->d_has_neg_pto.set( false );
-  }
-}
-
-void TheorySep::addPto( HeapAssertInfo * ei, Node ei_n, Node p, bool polarity ) {
-  Trace("sep-pto") << "Add pto " << p << ", pol = " << polarity << " to eqc " << ei_n << std::endl;
-  if( !ei->d_pto.get().isNull() ){
-    if( polarity ){
-      Trace("sep-pto-debug") << "...eqc " << ei_n << " already has pto " << ei->d_pto.get() << ", merge." << std::endl;
-      mergePto( ei->d_pto.get(), p );
-    }else{
-      Node pb = ei->d_pto.get();
-      Trace("sep-pto") << "Process positive/negated pto " << " " << pb << " " << p << std::endl;
-      Assert(pb.getKind() == kind::SEP_LABEL
-             && pb[0].getKind() == kind::SEP_PTO);
-      Assert(p.getKind() == kind::SEP_LABEL && p[0].getKind() == kind::SEP_PTO);
-      Assert(areEqual(pb[1], p[1]));
-      std::vector< Node > exp;
-      if( pb[1]!=p[1] ){
-        // if( pb[1].getKind()==kind::SET_SINGLETON &&
-        // p[1].getKind()==kind::SET_SINGLETON ){
-        //  exp.push_back( pb[1][0].eqNode( p[1][0] ) );
-        //}else{
-        exp.push_back( pb[1].eqNode( p[1] ) );
-        //}
+      else if (polarity != pol)
+      {
+        // a positive and negative pto
+        if (!areDisequal(pval, qval))
+        {
+          std::vector<Node> exp;
+          if (plbl != qlbl)
+          {
+            Assert(areEqual(plbl, qlbl));
+            exp.push_back(plbl.eqNode(qlbl));
+          }
+          Node pos = polarity ? p : q;
+          Node neg = polarity ? q : p;
+          exp.push_back(pos);
+          exp.push_back(neg.notNode());
+          std::vector<Node> conc;
+          if (pval != qval)
+          {
+            conc.push_back(pval.eqNode(qval).notNode());
+          }
+          Node concn = nm->mkOr(conc);
+          Trace("sep-pto") << "prop neg/pos: " << concn << " by " << exp
+                           << std::endl;
+          // propagation (pto x y) ^ ~(pto z w) ^ x = z => y != w
```

Comment:
Same issue here. The comment uses  equality between locations `x, z` but the code uses equality between sets `plbl, qlbl`